api/librats_verify_evidence.c

```diff
@@ -10,140 +11,199 @@
 
-rats_verifier_err_t librats_verify_evidence(attestation_evidence_t *evidence, const uint8_t *hash,
-					    claim_t **claims, size_t *claims_length)
+rats_verifier_err_t librats_verify_evidence(uint8_t *evidence_buffer, size_t evidence_buffer_size,
+					    uint8_t *endorsements_buffer,
+					    size_t endorsements_buffer_size, claim_t **claims,
+					    size_t *claims_length)
 {
-	uint32_t hash_len = 32;
 	rats_core_context_t ctx;
 	rats_conf_t conf;
+	uint32_t hash_len = SHA256_DIGEST_LENGTH;
+	uint8_t hash[SHA256_DIGEST_LENGTH];
+	attestation_evidence_t evidence;
+	uint8_t *p;
+	uint8_t *custom_claims_buffer;
+	size_t custom_claims_buffer_size;
+	claim_t *custom_claims = NULL;
+	size_t custom_claims_length;
+	claim_t *builtin_claims = NULL;
+	size_t builtin_claims_length;
+	claim_t *mixed_claims = NULL;
+
+	rats_verifier_err_t ret = RATS_VERIFIER_ERR_NONE;
 
 	conf.api_version = RATS_API_VERSION_DEFAULT;
 	conf.log_level = RATS_LOG_LEVEL_DEFAULT;
-	memcpy(conf.verifier_type, evidence->type, sizeof(conf.verifier_type));
 
+	/* Deserialize evidence */
+	p = evidence_buffer;
+	ret = RATS_VERIFIER_ERR_INVALID;
+	if (deserialize_evidence(&evidence, &p) != RATS_ERR_NONE)
+		goto err;
+
+	/* Calculate hash of custom_claims_buffer */
+	custom_claims_buffer = p;
+	custom_claims_buffer_size = evidence_buffer_size - (custom_claims_buffer - evidence_buffer);
+
+	SHA256(custom_claims_buffer, custom_claims_buffer_size, hash);
+
+	ret = RATS_VERIFIER_ERR_INIT;
 	if (rats_verify_init(&conf, &ctx) != RATS_VERIFIER_ERR_NONE)
-		return RATS_VERIFIER_ERR_INIT;
+		goto err;
+
+	ret = ctx.verifier->opts->verify_evidence(ctx.verifier, &evidence, hash, hash_len,
+						  &builtin_claims, &builtin_claims_length);
+	if (ret != RATS_VERIFIER_ERR_NONE)
+		goto err;
+
+	ret = RATS_VERIFIER_ERR_INVALID;
+	if (deserialize_claims_list(custom_claims_buffer, custom_claims_buffer_size, &custom_claims,
+				    &custom_claims_length) != RATS_ERR_NONE)
+		goto err;
 
-	rats_verifier_err_t err = ctx.verifier->opts->verify_evidence(
-		ctx.verifier, evidence, hash, hash_len, claims, claims_length);
+	mixed_claims = builtin_claims;
+	ret = RATS_VERIFIER_ERR_NO_MEM;
+	if (!realloc(mixed_claims,
+		     sizeof(claim_t) * (builtin_claims_length + custom_claims_length)))
+		goto err;
+	builtin_claims = NULL;
+
+	memcpy(mixed_claims + builtin_claims_length, custom_claims,
+	       sizeof(claim_t) * custom_claims_length);
+	free(custom_claims); /* Just free claims array and keep its content  */
+	custom_claims = NULL;
+	*claims = mixed_claims;
+	*claims_length = builtin_claims_length + custom_claims_length;
+	ret = RATS_VERIFIER_ERR_NONE;
+
+err:
+	if (custom_claims)
+		free_claims_list(custom_claims, custom_claims_length);
+
+	if (builtin_claims)
+		free_claims_list(builtin_claims, builtin_claims_length);
 
 	if (ctx.verifier->opts->cleanup(ctx.verifier) != RATS_VERIFIER_ERR_NONE) {
 		RATS_ERR("failed to clean up verifier\n");
+		return RATS_VERIFIER_ERR_INVALID;
```

Comment:
I think `*_ERR_INVALID` represents most INVALID cases, such as an invalid evidence, and also, wrong parameters, etc. Is that right?

Here I prefer to delete `return RATS_VERIFIER_ERR_INVALID;`, because it overwrites the value of `ret`.